src/common/bfloat16.hpp

```diff
@@ -22,20 +22,49 @@
 #include <cstdint>
 #include <cstdlib>
 #include <limits>
+#include <type_traits>
+
+#include "common/bit_cast.hpp"
 
 #include "dnnl.h"
 
 namespace dnnl {
 namespace impl {
 
 struct bfloat16_t {
+private:
+    // Converts the 32 bits of a normal float or zero to the bits of a bfloat16.
+    static constexpr uint16_t convert_bits_of_normal_or_zero(
+            const uint32_t bits) {
+        return uint32_t {
+                       bits + uint32_t {0x7FFFU + (uint32_t {bits >> 16} & 1U)}}
+        >> 16;
+    }
+
+public:
```

Comment:
really minor nit: could you please put the `private` towards the end of the struct, as it is more of an implementation detail. Also, if you don't mind, please drop the `public:` as it goes to the top of the struct, which is already public by default.

Reply:
> could you please put the `private` towards the end of the struct...

No problem @emfomenk, addressed by the amend/force-push that I just did : https://github.com/oneapi-src/oneDNN/compare/6c7d534caf60b0f5c106396e3048332eace152de..51638e755c39a6c0a2e15cf340101b67e0d9ee49

The amend also adds an extra unit test, which aims to check (in an indirect manner) that SFINAE prevents the new integer-specific member function templates from being called for a `double` argument:  `test_bfloat16_denorm_f32.BitsFromDoubleSameAsFromFloat`.

As submitter of the PR, I have to admit it has a little weakness. The converting constructor template silently allows conversions that may in fact be narrowing (potentially lossy). But personally I find this acceptable, given the fact that the existing `bfloat16_t(float)` converting constructor is already "potentially lossy" (from 32-bit to 16-bit). So now I do think this PR is ready anyway  😃

Reply:
> The converting constructor template silently allows conversions that may in fact be narrowing (potentially lossy).

Could you please elaborate on this a little more? As far as I can tell, SFINAE will prevent from using a new ctor with non-integer argument. What could go wrong?

Reply:
> As far as I can tell, SFINAE will prevent from using a new ctor with non-integer argument. What could go wrong?

@emfomenk  Indeed, as far as I can see, the new member function templates only get instantiated for integer argument types. That should be fine. However, many 32-bit integer values cannot be represented _exactly_ by a bfloat16 (for example, `257`), so that could go wrong. The implicit conversion from integer to bfloat16 could be lossy. (Just like conversion from 32-bit `float` to bfloat16 could be lossy.) Now before this PR, such conversion did already compile, but at least, you might get a warning:
```
struct my_bfloat16_t {
  my_bfloat16_t(float);
};

bool my_func(my_bfloat16_t bf16);
int i;
bool b = my_func(i); // warning C4244: 'argument': conversion from 'int' to 'float', possible loss of data
```
See https://godbolt.org/z/Gca6ob

This PR will avoid such a warning. Is that a good thing or a bad thing? At least, it's no show stopper to me!

Reply:
Now I see what you meant. Thanks for the explanation. We are fine with the change. In fact, given how often the library moves the data back and force (because it has too) we have to work around the warning by using explicit casting, which is oK but also bloats the code (and makes it sometime less readable). We might've even disabled the warning altogether if the project were a standalone application; but given it is a library that is often embedded to other applications we try minimizing the amount of warnings (even if they are false-postive) caused by our source code.

As you said, those who plan using `bfloat16_t` already know about lossy properties of the type, so that shouldn't be a problem.